drivers/video/video_mcux_elcdif.c

```diff
@@ -0,0 +1,425 @@
+	struct video_buffer *next_vbuf;
+	u32_t status;
+
+	status = ELCDIF_GetInterruptStatus(config->base);
+	ELCDIF_ClearInterruptStatus(config->base, status);
+
+	if (!(status & kELCDIF_CurFrameDone)) {
+		return;
+	}
+
+	if (data->current_vbuf == NULL)
+		goto next;
+
+	if (data->current_vbuf != data->next_vbuf) {
+		/* current buffer done, release it */
+		k_fifo_put(&data->fifo_out, data->current_vbuf);
+#ifdef CONFIG_POLL
+		if (data->signal) { /* Trigger frame done event */
+			k_poll_signal_raise(data->signal, VIDEO_BUF_DONE);
+		}
+#endif
+	}
+
+next:
+	/* next buffer is transfered now (possibly the same) */
+	data->current_vbuf = data->next_vbuf;
+
+	/* Do we have any new buffer for next ? */
+	next_vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT);
+	if (next_vbuf) {
+		data->next_vbuf = next_vbuf;
+		ELCDIF_SetNextBufferAddr(config->base,
+					 (uint32_t)data->current_vbuf->buffer);
+	}
+}
+
+static int video_mcux_elcdif_init(struct device *dev)
+{
+	const struct video_mcux_elcdif_config *config = dev->config->config_info;
+	struct video_mcux_elcdif_data *data = dev->driver_data;
+
+	k_fifo_init(&data->fifo_in);
+	k_fifo_init(&data->fifo_out);
+
+	ELCDIF_RgbModeGetDefaultConfig(&data->elcdif_config);
+
+	/* Some manual tunning for now */
+	data->elcdif_config.polarityFlags =
+			 kELCDIF_DataEnableActiveHigh |
+			 kELCDIF_VsyncActiveLow |
+			 kELCDIF_HsyncActiveLow |
+			 kELCDIF_DriveDataOnRisingClkEdge;
+	data->elcdif_config.dataBus = kELCDIF_DataBus16Bit;
+	data->elcdif_config.pixelFormat = kELCDIF_PixelFormatRGB565;
+
+	/* check if there is any attached lcd device */
+	if (config->lcd_label) {
+		data->lcd_dev = device_get_binding(config->lcd_label);
+		if (data->lcd_dev == NULL) {
+			LOG_ERR("Unable to get lcd device %s\n",
+				config->lcd_label);
+			return -ENODEV;
+		}
+	}
+
+	/* todo remove */
+	ELCDIF_RgbModeInit(config->base, &data->elcdif_config);
+
+#ifdef CONFIG_VIDEO_DISPLAY
+	display_video_init(dev);
+#endif
+
+	return 0;
+}
+
+#ifdef CONFIG_POLL
+static int video_mcux_elcdif_set_signal(struct device *dev,
+				     enum video_endpoint_id ep,
+				     struct k_poll_signal *signal)
+{
+	struct video_mcux_elcdif_data *data = dev->driver_data;
+
+	if (data->signal && signal != NULL) {
+		return -EALREADY;
+	}
+
+	data->signal = signal;
+
+	return 0;
+}
+#endif
+
+static const struct video_driver_api video_mcux_elcdif_driver_api = {
+	.set_format = video_mcux_elcdif_set_fmt,
+	.get_format = video_mcux_elcdif_get_fmt,
+	.stream_start = video_mcux_elcdif_stream_start,
+	.stream_stop = video_mcux_elcdif_stream_stop,
+	.flush = video_mcux_elcdif_flush,
+	.enqueue = video_mcux_elcdif_enqueue,
+	.dequeue = video_mcux_elcdif_dequeue,
+	.set_ctrl = video_mcux_elcdif_set_ctrl,
+	.get_ctrl = video_mcux_elcdif_get_ctrl,
+	.get_caps = video_mcux_elcdif_get_caps,
+#ifdef CONFIG_POLL
+	.set_signal = video_mcux_elcdif_set_signal,
+#endif
+#ifdef CONFIG_VIDEO_DISPLAY
+	.display_drv_api.blanking_on = display_video_blanking_on,
```

Comment:
I took a closer look at this commit and wonder what it will be? The ELCDIF driver has been moved to video although it is a display driver, then there is new display_video driver (which exposes self to everything) to provide "backward display API compatibility". Will display API be deprecated? You need a demultiplexer for the video source or drain? then please implement it.

Reply:
> I took a closer look at this commit and wonder what it will be? The ELCDIF driver has been moved to video although it is a display driver

Well, The video API is more generic than display, it's any device that can generate or consume video buffers (camera, encoder, display...), this API mainly allows to create zero-copy video pipelines. ELCDIF is not really a display but an 'LCD interface device' which can be connected to a 'LCD display device' and thus can be implemented as a generic video driver (output video device).

> then there is new display_video driver (which exposes self to everything) to provide "backward display API compatibility".

Exactly, this avoids code duplication. Moreover display is currently requested because LVGL relies on Display API and cannot interact directly with Video API.

> Will display API be deprecated? You need a demultiplexer for the video source or drain? then please implement it.

Not for now, but I need to discuss it with @vanwinkeljan to check if it's possible to converge at some point. Display API allows to write specific region of a frame buffer, video only takes full frame buffers pointers.

Reply:
> > I took a closer look at this commit and wonder what it will be? The ELCDIF driver has been moved to video although it is a display driver
> 
> Well, The video API is more generic than display, it's any device that can generate or consume video buffers (camera, encoder, display...), this API mainly allows to create zero-copy video pipelines. ELCDIF is not really a display but an 'LCD interface device' which can be connected to a 'LCD display device' and thus can be implemented as a generic video driver (output video device).


No, it is not and definitely not suitable for displays. ELCDIF is a generic display controller, and what we usually have are display controller drivers in `drivers/display`. ELCDIF driver should stay there.

> 
> > then there is new display_video driver (which exposes self to everything) to provide "backward display API compatibility".
> 
> Exactly, this avoids code duplication. Moreover display is currently requested because LVGL relies on Display API and cannot interact directly with Video API.


I do not see any duplication if you would use display API.

> 
> > Will display API be deprecated? You need a demultiplexer for the video source or drain? then please implement it.
> 
> Not for now, but I need to discuss it with @vanwinkeljan to check if it's possible to converge at some point. Display API allows to write specific region of a frame buffer, video only takes full frame buffers pointers.

Display API does not forbid you to write a full frame at once. 
I am not entire happy with the display API but I am sure it should not be mixed with video API. I will block this PR. APIs should be kept simple and do what they were created for.

Reply:
> > > I took a closer look at this commit and wonder what it will be? The ELCDIF driver has been moved to video although it is a display driver
> > 
> > 
> > Well, The video API is more generic than display, it's any device that can generate or consume video buffers (camera, encoder, display...), this API mainly allows to create zero-copy video pipelines. ELCDIF is not really a display but an 'LCD interface device' which can be connected to a 'LCD display device' and thus can be implemented as a generic video driver (output video device).
> 
> No, it is not and definitely not suitable for displays. ELCDIF is a generic display controller, and what we usually have are display controller drivers in `drivers/display`. ELCDIF driver should stay there.
> 
> > > then there is new display_video driver (which exposes self to everything) to provide "backward display API compatibility".
> > 
> > 
> > Exactly, this avoids code duplication. Moreover display is currently requested because LVGL relies on Display API and cannot interact directly with Video API.
> 
> I do not see any duplication if you would use display API.
> 
> > > Will display API be deprecated? You need a demultiplexer for the video source or drain? then please implement it.
> > 
> > 
> > Not for now, but I need to discuss it with @vanwinkeljan to check if it's possible to converge at some point. Display API allows to write specific region of a frame buffer, video only takes full frame buffers pointers.
> 
> Display API does not forbid you to write a full frame at once.
> I am not entire happy with the display API but I am sure it should not be mixed with video API. I will block this PR. APIs should be kept simple and do what they were created for.

I do not understand, I do not modify any API here, I moved the ELCDIF as a simpler video driver which can implement either video API or display API.

Reply:
> I do not understand, I do not modify any API here, I moved the ELCDIF as a simpler video driver which can implement either video API or display API. The current display API allows to write frame (which implies copy), that's true but not to just display frame (zero copy). Video API is just a device type agnostic layer allowing to queue dequeue video buffer from/to any kind of video device (including display). If it's jsut about folder we can keep it under drivers/display/.

Do not move ELCDIF driver to video, do not use video API as display API. Implement a demuxer to pass the data from a video source to a device like display controller.

Reply:
> > I do not understand, I do not modify any API here, I moved the ELCDIF as a simpler video driver which can implement either video API or display API. The current display API allows to write frame (which implies copy), that's true but not to just display frame (zero copy). Video API is just a device type agnostic layer allowing to queue dequeue video buffer from/to any kind of video device (including display). If it's jsut about folder we can keep it under drivers/display/.
> 
> Do not move ELCDIF driver to video, do not use video API as display API. Implement a demuxer to pass the data from a video source to a device like display controller.

This is something to discuss, since I tend to disagree, if requested I'll create a separated video driver for ELCDIF, indeed this is a perfect candidate for video interface API since it just pushes data to a display and thus can be abstracted by an output device. Display devices can be quite complicated and controlled via an additional interface like I2C to configure/retrieve resolution, format... I don't see why adding a video to display 'demuxer' would be better than the display to video one as proposed in this PR (video is more generic than display). I just want to keep this simple and efficient without breaking the current display API/code, not make it just 'work' via overkill layers.

Reply:
> This is something to discuss, since I tend to disagree, if requested I'll create a separated video driver for ELCDIF, indeed this is a perfect candidate for video interface API since it just pushes data to a display and thus can be abstracted by an output device.

Again, do not move the display controller driver to video, use display API for display controller please. We usually push the data to the displays and use display API. 

>Display devices can be quite complicated and controlled via an additional interface like I2C to configure/retrieve resolution, format... 

Video devices too, so what?

>I don't see why adding a video to display 'demuxer' would be better than the display to video one as proposed in this PR (video is more generic than display). I just want to keep this simple and efficient without breaking the current display API/code, not make it just 'work' via overkill layers.

You do not have to break anything, just read the video data and write it to the display controller using display API. With the proper adaption layer you can use any display controller in the tree. Like you used socket API in the tcpserversink sample and did not tried to hack a video API - mcux ethernet API video driver.

Reply:
> > This is something to discuss, since I tend to disagree, if requested I'll create a separated video driver for ELCDIF, indeed this is a perfect candidate for video interface API since it just pushes data to a display and thus can be abstracted by an output device.
> 
> Again, do not move the display controller driver to video, use display API for display controller please. We usually push the data to the displays and use display API.

> > I don't see why adding a video to display 'demuxer' would be better than the display to video one as proposed in this PR (video is more generic than display). I just want to keep this simple and efficient without breaking the current display API/code, not make it just 'work' via overkill layers.
> 
> You do not have to break anything, just read the video data and write it to the display controller using display API. With the proper adaption layer you can use any display controller in the tree. Like you used socket API in the tcpserversink sample and did not tried to hack a video API - mcux ethernet API video driver.

It is indeed possible, but it's a poor solution, video API abstracts the components (camera, encoder, display) in a unified way, allowing user to use a simple common interface to retrieve and align/configure formats as well as to deal with buffers (cf capture_and_display sample). In that case, using display adds unnecessary complexity (different structures, functions...), much bigger memory footprint (display has two private dedicated buffers), latency/cpu-cycles (display implies copying buffer)... Moreover the current proposal is more flexible since display interface (ELCDIF) and display (LCD) are two separated devices, allowing to 'add/connect' any device to the interface without modifying the interface driver itself (dedicated LCD driver). That why I push for this unified solution. I can not force anything, but I don't get why it would not be suitable for zephyr to have such support, knowing that it still backward compatible with Display API.

Reply:
> much bigger memory footprint

This is just due to the fact you are supporting partial buffer writes in case you would enforce writing the whole buffer at once this buffers are not needed. Alternatively you could deque all video buffers and write the data to them and enqueue them.

> latency/cpu-cycles

This is the price you pay for using/supporting `display_write` API, the display API also provides an option to get the framebuffer directly `display_get_framebuffer` and perform actions directly on the framebuffer. Actually it is a bit strange that the ELCDIF display driver never implemented this support but I think this was more a time constraint then a technical issue. (Possibly we need to extend the display API with `display_put_framebuffer`).

> Moreover the current proposal is more flexible since display interface (ELCDIF) and display (LCD) are two separated devices

I like this approach but this has nothing to do with video API and could have been achieved with the display API.

@jfischer-phytec-iot I have no problem in deprecating the display API and go to one unified API. But it is a hard requirement to support partial buffers and the unified API should have minimal dependencies to other subsystems.

Reply:
> > much bigger memory footprint
> 
> This is just due to the fact you are supporting partial buffer writes in case you would enforce writing the whole buffer at once this buffers are not needed. Alternatively you could deque all video buffers and write the data to them and enqueue them.
> 
> > latency/cpu-cycles
> 
> This is the price you pay for using/supporting `display_write` API, the display API also provides an option to get the framebuffer directly `display_get_framebuffer` and perform actions directly on the framebuffer. Actually it is a bit strange that the ELCDIF display driver never implemented this support but I think this was more a time constraint then a technical issue. (Possibly we need to extend the display API with `display_put_framebuffer`).

Yes, since we need a way to dequeue/enqueue (get/put). An other point is that getting buffer is a blocking function in display API, so it could request its own dedicated thread to perform concurrent operations (getting camera buffer, sw image processing).

> > Moreover the current proposal is more flexible since display interface (ELCDIF) and display (LCD) are two separated devices
> 
> I like this approach but this has nothing to do with video API and could have been achieved with the display API.

Correct.

> @jfischer-phytec-iot I have no problem in deprecating the display API and go to one unified API. But it is a hard requirement to support partial buffers and the unified API should have minimal dependencies to other subsystems.

well, that why I went with this backward compatible solution, display API can not really be deprecated for now, but I would be happy to work with you on unifying the APIs (probably progressively, adding specific video controls for display, LVGL support, etc), maybe starting by using common formats, pixel code, buffer formats... Currently, partial write is possible with video API since write can be done directly in the buffer before enqueuing it for output. At the end I agree it's is possible to improve the display API, but I'm much in favor of creating one unified API and make things easier for user (like it can be with Linux Video APIs, v4l2, gstreamer...).

Reply:
So really slow down. You do not have a solution here, it is a bad hack nothing more. Even with an one common overloaded API you have to implement a system between data source and data drain because they would not always have the same format like resolution and pixel format. Your video API is far from the point to be proven even for video sources, I can only see the CSI interface and one driver for the sensors. Display API is not perfect but much more mature. Again, the ELCDIF is a display controller and the driver should use display API. If you want to use ELCDIF or an other display controller as a drain for the data from a video source then add a sample/subsystem/layer for that, like it was done for tcpserversink!

Reply:
I understand the maturity point, though I don't want to replace the display API here, but make the ELCDIF support both display API and video API, which normally does not cause any regression and is, in my opinion, a good compromise. This makes the sample easier to understand, format configuration/alignment straightforward and tested with compatible devices (well, just the ELCDIF + RK043FN02H for now) since I'm not sure any display can support 'video'. I can take a step back and go with the display API solution which would certainly work. But then I don't know what to do after that, should we create a dedicate API for camera, encoder, decoder, display, video output... or trying to move on a unified version. Don't want an endless ping-pong here, I know we don't agree on the way to go, just trying to find a way to move forward on this PR and on next steps.